### PR TITLE
feat: support rate limiter based on db and partition levels

### PR DESCRIPTION
**Pull request by SimFG**
issue: https://github.com/milvus-io/milvus/issues/30577
co-author: @jaime0815

### CODE REVIEW

File: internal/rootcoord/quota_center.go [hunk 1]
```diff
@@ -844,118 +1248,214 @@ func (q *QuotaCenter) getCollectionLimitProperties(collection int64) map[string]
...
-		return
+		return nil
 	}
-	collections := typeutil.NewUniqueSet()
+
+	denyCluster := false
 	totalDiskQuota := Params.QuotaConfig.DiskQuota.GetAsFloat()
+	total := q.dataCoordMetrics.TotalBinlogSize
+	if float64(total) >= totalDiskQuota {
+		denyCluster = true
```
Comment (jaime0815) [inline internal/rootcoord/quota_center.go:1265]: force deny writing for cluster level and return directly

File: internal/rootcoord/quota_center.go [hunk 2]
```diff
@@ -844,118 +1248,214 @@ func (q *QuotaCenter) getCollectionLimitProperties(collection int64) map[string]
...
+	}
+
+	collectionDiskQuota := Params.QuotaConfig.DiskQuotaPerCollection.GetAsFloat()
+	dbSizeInfo := make(map[int64]int64)
+	collections := make([]int64, 0)
 	for collection, binlogSize := range q.dataCoordMetrics.CollectionBinlogSize {
 		collectionProps := q.getCollectionLimitProperties(collection)
-		colDiskQuota := getCollectionRateLimitConfig(collectionProps, common.CollectionDiskQuotaKey)
+		colDiskQuota := getRateLimitConfig(collectionProps, common.CollectionDiskQuotaKey, collectionDiskQuota)
```
Comment (jaime0815) [inline internal/rootcoord/quota_center.go:1273]: 1. Checking db size,  force deny writing them.
2. Check collection size in the next step and skip collection that in force deny abs.
3. checking partition size and skipping partitions that in force deny collections.

File: internal/util/quota/quota_constant.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+				internalpb.RateType_DMLDelete:   &quotaConfig.DMLMaxDeleteRatePerPartition,
+				internalpb.RateType_DMLBulkLoad: &quotaConfig.DMLMaxBulkLoadRatePerPartition,
+				internalpb.RateType_DQLSearch:   &quotaConfig.DQLMaxSearchRatePerPartition,
+				internalpb.RateType_DQLQuery:    &quotaConfig.DQLMaxQueryRatePerPartition,
+			},
+		}
+	})
+}
+
+func GetQuotaConfigMap(scope internalpb.RateScope, params *paramtable.ComponentParam) map[internalpb.RateType]*paramtable.ParamItem {
```
Comment (jaime0815) [inline internal/util/quota/quota_constant.go:88]: ensure params is a global object, or the second object will not be initialized. We'd better add the method comment.

File: internal/proxy/metrics_info.go [hunk 4]
```diff
@@ -50,12 +50,29 @@ func getQuotaMetrics() (*metricsinfo.ProxyQuotaMetrics, error) {
...
+				Label: s,
+				Rate:  f,
+			})
+		}
+	}
 	getRateMetric(internalpb.RateType_DMLInsert.String())
 	getRateMetric(internalpb.RateType_DMLUpsert.String())
 	getRateMetric(internalpb.RateType_DMLDelete.String())
 	getRateMetric(internalpb.RateType_DQLSearch.String())
+	getSubLabelRateMetric(internalpb.RateType_DQLSearch.String())
```
Comment (xiaofan-luan) [inline internal/proxy/metrics_info.go:71]: Do we really need SubLabelRateMetric?

Why not simply sum all collections from this database on rootcoord?

File: internal/proxy/simple_rate_limiter.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+			ret = collectionRateLimiters.Check(rt, n)
+			if ret != nil {
+				cancelAllLimiters()
+				return ret
+			}
+			doneLimiters = append(doneLimiters, collectionRateLimiters)
+		}
+	}
+
+	// 4. check partition level rate limits
```
Comment (xiaofan-luan) [inline internal/proxy/simple_rate_limiter.go:108]: is there a specific reason we need partiton level metrics?
Not sure how much time it cost by simply check all the limiters

File: internal/proxy/impl.go [hunk 6]
```diff
@@ -286,6 +288,7 @@ func (node *Proxy) DropDatabase(ctx context.Context, request *milvuspb.DropDatab
 	}
 
 	log.Info(rpcDone(method))
+	DeregisterSubLabel(ratelimitutil.GetDBSubLabel(request.GetDbName()))
```
[the remark this answers is not in the document]
Comment (SimFG) [reply internal/proxy/impl.go:291]: support to invalid the database meta cache, pr: #32078

File: internal/datacoord/broker/coordinator_broker.go [hunk 7]
```diff
@@ -39,6 +40,7 @@ type Broker interface {
 	ShowCollections(ctx context.Context, dbName string) (*milvuspb.ShowCollectionsResponse, error)
 	ListDatabases(ctx context.Context) (*milvuspb.ListDatabasesResponse, error)
 	HasCollection(ctx context.Context, collectionID int64) (bool, error)
+	GetDatabaseID(ctx context.Context, dbName string) (int64, error)
```
[the remark this answers is not in the document]
Comment (SimFG) [reply internal/datacoord/broker/coordinator_broker.go:43]: has added db id in DescribeCollectionResponse, pr: https://github.com/milvus-io/milvus/pull/32114

File: internal/proxy/metrics_info.go (hunk 4 above)
Comment (SimFG) [reply internal/proxy/metrics_info.go:71]: The current metric is the average of all collections, and detailed information about a collection cannot be obtained, so sub-labels are used to record more detailed information.

File: internal/proxy/simple_rate_limiter.go (hunk 5 above)
Comment (SimFG) [reply internal/proxy/simple_rate_limiter.go:108]: The partition here is the partition involved in the request, and the corresponding limiter is obtained directly through the map, so the check should not consume a lot of resources.

**Pull request merged**